Local network detection and IPv6 support

coolparadox opened the pull request:
New [test cases](/turbobytes/geoipdb/blob/detect_private_network/geoipdb_test.go#L350) to tell private network apart from other types of failure are working fine.

This adds new error values to AsnLookup() only -- there is no change to geoipdb interface. CNC builds successfully with this merge.

Fixes #8

Commit:
make cache concurrency safe fixes #24

coolparadox committed:
Remove test binary uploaded by mistake.

And protect against inadvertently commit results of 'go test -c' again.

References #8

coolparadox committed:
iputils/IsLocalIp() stub and test cases

References #8

coolparadox committed:
Implementation of original islocalip() algorithm from pulse/utils

An extended table of non global ranges of IP addresses is in place.
Originally used IP values pass in tests.

There is no possibility to tune or disable detection on the fly at the
time.

References #8

coolparadox committed:
Parse CIDRs of private networks on init.

References #8

coolparadox committed:
Handler methods return specific errors on malformed and IPv6 ip parameter.

References #8

coolparadox committed:
LookupAsn return specific error for private IPs

Added test cases as well for new LookupAsn errors

References #8

coolparadox committed:
Export IsIPv6()

References #8

coolparadox committed:
Added several new test cases for LookupAsn

References #8

coolparadox commented:
(Please hold review as it looks there are failing test cases in Travis, let me look)

coolparadox committed:
Fix test case for updated MaxMind database.

References #8

coolparadox commented:
All set, please proceed with review

coolparadox committed:
Silent warning about malformed ASN in Cymru lookup.

References #8

coolparadox committed:
Preliminary removal of IPv6NotSupportedError

References #9

coolparadox committed:
Query IPv6 Maxmind database

Also fix test cases for IPv6.

References #9

coolparadox committed:
Download IPv6 Maxmind database for travis tests

References #9

coolparadox commented:
IPv6 support in geoipdb is enabled:

```
$ http GET localhost:7778/asnlookup/ip/2001:4860:1004::876:102
HTTP/1.1 200 OK
Content-Encoding: gzip
Content-Length: 186
Content-Type: application/json; charset=UTF-8
Date: Tue, 13 Dec 2016 13:19:10 GMT

{
    "ip": "2001:4860:1004::876:102",
    "result": {
        "asndb": {
            "asn": "AS15169",
            "err": "ASN not found",
            "name": ""
        },
        "cymru": {
            "asn": "AS15169",
            "err": "",
            "name": "GOOGLE - Google Inc., US"
        },
        "geoipdb": {
            "asn": "AS15169",
            "err": "",
            "name": "Google Inc."
        },
        "ipinfo": {
            "asn": "AS15169",
            "err": "",
            "name": "Google Inc."
        },
        "maxmind": {
            "asn": "AS15169",
            "err": "",
            "name": "Google Inc."
        }
    }
}
```

Fixes #9

sajal commented on geoipdb.go:
```diff
@@ -111,15 +125,21 @@ func NewHandler(overrides *mgo.Collection, timeout time.Duration) (Handler, erro
 func (h Handler) LibGeoipLookup(ip string) (string, string) {
-	if ip == "" {
+	var name string
+	ip_, is4 := iputils.ParseIP(ip)
```
why are you using `ip_` ? I think underscore might confuse people thinking there is an ignored second return argument.

coolparadox replied:
Agreed; fixing...

coolparadox committed:
Don't use variable names with _ suffix

As it may confuse readers into thinking there may be ignored return
parameters.

References #25
